### DISCUSSION TITLE
Short Tutorials

### DISCUSSION THREAD

**Discussion by S2000Stefan** [terjeio/ioSender#197] [Show and tell]
Hello,

I plan to make some short tutorials to show the functionality of ioSender. I would like to make them available to terjeio,
for free of course, to make his work better known and more user-friendly.

Unfortunately, with the way I work, this will definitely take some time. So please don't expect so much from me. :)
Here is my first draft of how I imagine it, I am very open to suggestions and criticism. To my apology english is not my mother tongue as you can see.

https://www.youtube.com/watch?v=YFT425f7KBo
new revised version

Unfortunately, I don't have my own webspace and have to upload the video to youtube first. Either you solve this later in a different way or leave the videos on youtube because of the better distribution.

Now to the problem that I have noticed in my work.

If you make an entry in the probing tap under "Probing distances and speed", you can no longer activate the jogging button with the mouse. No problem without an entry.

**Comment by terjeio** [comment]
Great, and looks good!

As for making your tutorials available I suggest you create a playlist on your Youtube channel, I can then link to this from the _Help_ menu and the main Wiki page.

> If you make an entry in the probing tap under "Probing distances and speed", you can no longer activate the jogging button with the mouse. 

This is a bug, I'll look into that. Ctrl+shift seems to still work if the button is not pressed first.

**Reply by terjeio** [reply]
I have fixed the "bug", a side-effect from refactoring jogging, available in the [latest edge version](http://www.io-engineering.com/downloads).

**Reply by S2000Stefan** [reply]
Thanks for the very quick bug fix, could only test it on the desktop first, machine will follow for sure. :+1: 

If only I could create my tutorials as fast as you always fix the bugs. :)

O.k. I will create a playlist for "Tutorials ioSender" on Youtube and hopefully be able to deliver more soon.

**Comment by phil-barrett** [comment]
Stefan,  I would be willing to put up a page for your tutorials on grbl.org.  I won't host the videos but happy to help otherwise.

**Reply by S2000Stefan** [reply]
> Stefan, I would be willing to put up a page for your tutorials on grbl.org. I won't host the videos but happy to help otherwise.

That would be a fine thing. But please wait a little longer until I have some videos ready. 
As I said, it may take me a while at the moment.
 I will also revise the first one, I am not quite satisfied with it yet.

**Comment by phil-barrett** [comment]
I will await your word on it.

**Comment by S2000Stefan** [comment]
Second tutorial is online. 

This time I've taken on the edge finder tab. 
It's a bit longer now, so it no longer falls under the short category. :) 
Here is the link:

https://youtu.be/TJhH2P-i16M
new revised version

**Comment by phil-barrett** [comment]
Really good job!  I like how it flows and you show what you are doing.  Very clear.

A couple of suggestions:

Not everyone will understand that you are making electrical contact between the probe (rod in the spindle) and the metal workpiece. Maybe add a small description of how you connected the probe and gnd inputs of the BOB to the spindle and workpiece?

I do not think you need to apologize for the machine shake.  Or at least, just do it once.

A small issue - the workpiece moves slightly from the rapid contacts. Might not be worth mentioning.

The url in your link doesn't work for me. I had to go to youtube.com and search for lVl0JMBwd1I  It looks like you inserted a URL for github instead of youtube. This one works.
[https://www.youtube.com/watch?v=lVl0JMBwd1I](https://www.youtube.com/watch?v=lVl0JMBwd1I)

**Reply by S2000Stefan** [reply]
Thank you Phil for your constructive criticism, I will try to improve some of the points you mentioned in the next versions.

> A small issue - the workpiece moves slightly from the rapid contacts. Might not be worth mentioning.

I also noticed this after looking at a large monitor. I will try to fix the workpiece better instead of simply placing it on the machine bed.

Will fix the link. But it's more github than me. :)

**Reply by terjeio** [reply]
Again, great work. I like how you illustrate how settings map back to the workpiece.
Please create a playlist so I can link to it.

> I do not think you need to apologize for the machine shake. Or at least, just do it once.

IMO not do it at all.

**Reply by S2000Stefan** [reply]
@terjeio
I have removed the repetitions from the above video and updated the link.
Also I have created the playlist, the link to it is: https://youtube.com/playlist?list=PLnSV6o2cRxM5mQQe4ec5cS2J8jBsEciY3
Keep in mind that there are only two videos online at the moment.
I hope there will be a few more soon.

**Comment by S2000Stefan** [comment]
@terjeio 
Today I wanted to start a tutorial about the tool change. 
I haven't used it for a very long time and have to say that the process is disturbed. 
After the first tool change, the machine moves back to position start as usual, the spindle moves to start position and then nothing happens. Here is a screenshot when it happens:
![IMG_20220223_153511_git](https://user-images.githubusercontent.com/35189051/155345747-4ceffbcd-e2c9-4bbf-a0c5-17911d251f9d.jpg)
Can you please take a look at it when you have time?
The line M3S7000 is not executed. The spindle does not accelerate to 7000 rpm.

**Comment by andrewmarles** [comment]
What build of GRBLHAL are you running?  This is very likely https://github.com/terjeio/ioSender/issues/195

Updating fixed it for me.

**Reply by S2000Stefan** [reply]
Yes, that could be a possibility, my grblHAL version is certainly older than 23 days.
I will update and try again.
Thank you.

**Comment by S2000Stefan** [comment]
Unfortunately, this time it's only a quick tip, it's about creating macros. The little helpers. :)

During my work I noticed that you can't call up a created macro directly with a function key. Is that the way it should be?

Also, ioSender crashed when I entered my Z probe macro and wanted to call it up. I think the error message was "Unhandled exception occured: The I/O operation......".
But I am not sure anymore

Am I writing this here or is it something to do with the screen recording? My milling computer really goes to pieces when a milling programme is running and I'm doing a screen recording at the same time.

Ah yes the link is: https://youtu.be/sj-8fS18VSA

**Reply by terjeio** [reply]
> During my work I noticed that you can't call up a created macro directly with a function key. Is that the way it should be?

No, and strange as it works for me.

> Also, ioSender crashed when I entered my Z probe macro and wanted to call it up. I think the error message was "Unhandled exception occured: The I/O operation......".

Did this only happen once?

> Am I writing this here or is it something to do with the screen recording? My milling computer really goes to pieces when a milling programme is running and I'm doing a screen recording at the same time.

Open the task manager and check the computer load. If too high this could be the cause.

---

A new feature is that [expressions](http://linuxcnc.org/docs/2.7/html/gcode/overview.html#gcode:expressions) can be used in macros:

![image](https://user-images.githubusercontent.com/20260062/156032402-85883e2d-dffc-4944-95b9-86b78de81b76.png)

Current machine [parameters](http://linuxcnc.org/docs/2.7/html/gcode/overview.html#_parameters) can be referenced, including the latest probing position.
I have not added support for conditional execution (if... then... else...), I may do so later. E.g. conditional execution could be used for checking if a probe action was successful or not and take appropriate action...

**Reply by S2000Stefan** [reply]
@terjeio 
>> During my work I noticed that you can't call up a created macro directly with a function key. Is that the way it should be?

>No, and strange as it works for me.

>>Also, ioSender crashed when I entered my Z probe macro and wanted to call it up. I think the error message was "Unhandled exception >>occured: The I/O operation......".

>Did this only happen once?

I will test the whole thing again tomorrow and report back, but I think it definitely had something to do with the screen recording.
I did this twice with the macro and ioSender crashed twice, but I will test it again tomorrow.

**Reply by S2000Stefan** [reply]
@terjeio 

Test result: 

The function keys work when you first click in the ioSender window. This is an operating error on my part.

I can repeat the second error with the crash. Enter the first two lines of my macro, close the macro window and then click directly afterwards with the mouse on the icon to execute. Then the crash always occurs, but if you execute the macro with the function key, everything is fine. 

Here are the two lines:
G91 G38.2 Z-10 F80
G90 G10 L20 P0 Z19.45

**Reply by terjeio** [reply]
@S2000Stefan I have just uploaded a new edge version with a fix.

**Reply by S2000Stefan** [reply]
@terjeio 
Is it possible that you are providing version 2.0.37p5 on your server and not p6? 
The macro issue is still present. 
The version number is also displayed as 2.0.37p5.
![p5](https://user-images.githubusercontent.com/35189051/156422504-e593c0ce-cf2a-4a90-abd0-0900dde51843.jpg)

**Reply by terjeio** [reply]
Oops - my bad. I have just uploaded the correct version. Sorry for this.

**Reply by S2000Stefan** [reply]
You don't have to apologise.
But I can only test again on Friday at the earliest.

**Comment by phil-barrett** [comment]
Nice!  The short ones are good.  I will reference this in the ioSender tour unless I hear otherwise.

**Comment by phil-barrett** [comment]
Sorry to mention this here but did you intend to have an infinite energy video in your playlist?  Just curious because it seems perhaps a bit out of place? 
![image](https://user-images.githubusercontent.com/16570958/156101239-2fcc2855-3c9a-4313-8ecb-f03df98f83e9.png)

**Comment by S2000Stefan** [comment]
@phil-barrett 
> Sorry to mention this here but did you intend to have an infinite energy video in your playlist? Just curious because it seems perhaps a bit out of place?

Oh oh, I don't know how the video got in there. It's not even mine, I had watched it a few days ago??? 
I removed it immediately.

Thanks Phil for reporting it to me, I certainly wouldn't have noticed.

**Comment by phil-barrett** [comment]
I didn't figure you for a perpetual motion kind of guy!  I would have contacted you more discreetly but didn't see a way.  It would be nice if github had some sort of private message feature.

**Reply by S2000Stefan** [reply]
> I didn't figure you for a perpetual motion kind of guy! I would have contacted you more discreetly but didn't see a way. It would be nice if github had some sort of private message feature.

I'm not a perpetual motion machine guy, but I find the idea interesting ;-)
But I would like to know how the people in the video do the magic trick.

It was right to point this out to me, the video had no place there and as I said, I don't know how it got in.

**Comment by S2000Stefan** [comment]
A new tutorial is online.

Semi automatic tool change @G59.3

In this tutorial I would like to show you how I use 
ioSender 2.0.37 in conjunction with grblHAL 
to perform a semi-automatic tool change 
at position G59.3.
Link: https://youtu.be/nU9RS5zZdo4
